File: library/src/main/java/com/yausername/youtubedl_android/YoutubeDLOptions.java
package com.yausername.youtubedl_android;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class YoutubeDLOptions {

    private final Map<String, List<String>> options = new LinkedHashMap<>();

    public YoutubeDLOptions addOption(@NonNull String option, @NonNull String argument) {
        if (!options.containsKey(option)) {
            List<String> arguments = new ArrayList<>();
            arguments.add(argument);
            options.put(option, arguments);
        } else {
            options.get(option).add(argument);
        }
        return this;
    }

    public YoutubeDLOptions addOption(@NonNull String option, @NonNull Number argument) {
        if (!options.containsKey(option)) {
            List<String> arguments = new ArrayList<>();
            arguments.add(argument.toString());
            options.put(option, arguments);
        } else {
            options.get(option).add(argument.toString());
        }
        return this;
    }

    public YoutubeDLOptions addOption(String option) {
        if (!options.containsKey(option)) {
            List<String> arguments = new ArrayList<>();
            arguments.add("");
            options.put(option, arguments);
        } else {
            options.get(option).add("");
        }
        return this;
    }

    public String getArgument(String option) {
        if (!options.containsKey(option))
            return null;
        String argument = options.get(option).get(0);
        if (argument.isEmpty())
            return null;
        else return argument;
    }

    public List<String> getArguments(String option) {
        if (!options.containsKey(option))
            return null;
        return options.get(option);
    }

    public boolean hasOption(String option) {
        return options.containsKey(option);
    }

    public List<String> buildOptions() {
        List<String> commandList = new ArrayList<>();
        for (Map.Entry<String, List<String>> entry : options.entrySet()) {
            String option = entry.getKey();
            for (String argument : entry.getValue()) {
                commandList.add(option);
                if (!argument.isEmpty())
                    commandList.add(argument);
            }
        }
        return commandList;
    }

}

File: library/src/main/java/com/yausername/youtubedl_android/YoutubeDLOptions.kt
package com.yausername.youtubedl_android

import kotlin.collections.ArrayList
import kotlin.collections.LinkedHashMap
import kotlin.collections.List
import kotlin.collections.MutableList
import kotlin.collections.MutableMap
import kotlin.collections.component1
import kotlin.collections.component2
import kotlin.collections.iterator
import kotlin.collections.set

class YoutubeDLOptions {
    private val options: MutableMap<String, MutableList<String>> = LinkedHashMap()
    fun addOption(option: String, argument: String): YoutubeDLOptions {
        if (!options.containsKey(option)) {
            val arguments: MutableList<String> = ArrayList()
            arguments.add(argument)
            options[option] = arguments
        } else {
            options[option]!!.add(argument)
        }
        return this
    }

    fun addOption(option: String, argument: Number): YoutubeDLOptions {
        if (!options.containsKey(option)) {
            val arguments: MutableList<String> = ArrayList()
            arguments.add(argument.toString())
            options[option] = arguments
        } else {
            options[option]!!.add(argument.toString())
        }
        return this
    }

    fun addOption(option: String): YoutubeDLOptions {
        if (!options.containsKey(option)) {
            val arguments: MutableList<String> = ArrayList()
            arguments.add("")
            options[option] = arguments
        } else {
            options[option]!!.add("")
        }
        return this
    }

    fun getArgument(option: String): String? {
        if (!options.containsKey(option)) return null
        val argument = options[option]!![0]
        return argument.ifEmpty { null }
    }

    fun getArguments(option: String): List<String>? {
        return if (!options.containsKey(option)) null else options[option]
    }

    fun hasOption(option: String): Boolean {
        return options.containsKey(option)
    }

    fun buildOptions(): List<String> {
        val commandList: MutableList<String> = mutableListOf()
        for ((option, value) in options) {
            for (argument in value) {
                commandList.add(option)
                if (argument.isNotEmpty()) commandList.add(argument)
            }
        }
        return commandList
    }
}